src/library/scala/collection/IterableOnce.scala

```diff
@@ -767,6 +767,13 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     }
     else onEmpty
   }
+  // Invoke reduce instead of reduceLeftIterator from sum, product, min, max to use overridden reduce or reduceLeft.
+  // Take this path if knownSize == -1. It incurs exception overhead when empty.
+  private final def reduceFallback[B >: A](onEmpty: => B)(op: (B, B) => B): B =
+    try reduce(op)
+    catch {
+      case e: UnsupportedOperationException if e.getMessage == "empty.reduceLeft" => onEmpty
```

Comment:
Er, that does not seem like good design to me.

What if someone overrides `reduce` with a different error message? What if `op` itself throws an `UnsupportedOperationException`? Not to mention the performance cost of throwing and catching for something that is not an exception situation.

Reply:
Also considered just fixed exception object to throw to mitigate some of that.

The requirements in tension are "check isEmpty if knownSize is -1" but "without creating an extra iterator" and "let sum work with custom reduce".

The other option is make plain you must override sum etc. Or, make a template method for overriding. Of course it is brittle to let them override everything, so design is backed into a corner for extension.

Reply:
BigDecimal math context doesn't like to be folded. I don't have that old ticket handy.

Reply:
that is for sum, product.

Reply:
This can simmer awhile. lrytz will have a response (disgust). The other idea is just to revert sum and document the behavior (it uses isEmpty and your choice is knownSize or probe with iterator). Maybe that is most sane for these methods.

Reply:
Why not just `reduceOption(op).getOrElse(onEmpty)`?
`reduceOption` can be overridden. And as `reduce` is only "supported" (as the Exception tells us) for non-empty collections, `reduceOption` would imho be the better method to override anyway for collections that may be empty.

Reply:
Somewhere Ichoran suggested reduceOption. My revision will use isEmpty, on assumption that they should implement either knownSize or isEmpty. There is also the maxBy trick, a special accumulator. I had not considered just documenting that for this edge case, you can override reduceOption instead if necessary.